docs/service-add-ons.md

```diff
@@ -0,0 +1,170 @@
+
+# The service lists components not directly related to data pipelines, but used
+# by the service.
+service:
+  # add-ons lists the add-on components added to the service. They are started
+  # in the order presented below and stopped in the reverse order.
+  add-ons: [health-check, pprof, zpages]
+```
+
+The current command-line related flags will be removed as the components are 
+moved to the new model. Command-line flags will be limited to settings directly 
+affecting the service as a whole, e.g.: logging.
+
+The configuration base type won’t share any common fields. Existing ad-hoc 
+components use endpoint but it is possible to have some other add-ons that do 
+not use that configuration at all.
+
+The factory will follow the pattern established for pipeline configuration:
+
+```go
+// Factory is a factory interface for add-ons to the service.
+type Factory interface {
+    // Type gets the type of the add-on component created by this factory.
+    Type() string 
+
+    // CreateDefaultConfig creates the default configuration for the add-on.
+    CreateDefaultConfig() configmodels.AddOn 
+
+    // CustomUnmarshaler returns a custom unmarshaler for the configuration or nil if
+    // there is no need for custom unmarshaling. This is typically used if viper.Unmarshal()
+    // is not sufficient to unmarshal correctly.
+    CustomUnmarshaler(v *viper.Viper, viperKey string, intoCfg interface{}) CustomUnmarshaler 
+
+    // CreateAddOn creates a service add-on based on the given config.
+    CreateAddOn(logger *zap.Logger, cfg configmodels.AddOn) (Component, error)
+}
+```
+
+
+## Add-On Component Interface
+
+The interface defined below is the minimum required to keep same behavior for 
+ad-hoc components currently in use on the service:
+
+```go
+// Component is the interface for objects hosted by the OpenTelemetry Service that
+// doesn't participate directly on data pipelines but provide some functionality
+// to the service, examples: health check endpoint, z-pages, etc.
+type Component interface {
+    // Start the Component object hosted by the given host. At this point in the
+    // process life-cycle the receivers are not started and the host did not
+    // receive any data yet.
+    Start(host Host) error 
+
+    // Shutdown the Component instance. This happens after the pipelines were
+    // shutdown.
+    Shutdown() error
+}
+
+// PipelineWatcher is an extra interface for Components hosted by the OpenTelemetry
+// Service that is to be implemented by Components interested in changes to pipeline
+// states. Typically this will be used by Components that change their behavior if data is
+// being ingested or not, e.g.: a k8s readiness probe.
+type PipelineWatcher interface {
+    // Ready notifies the Component that all pipelines were built and the
+    // receivers were started, i.e.: the service is ready to receive data
+    // (notice that it may already have received data when this method is called).
```

Comment:
It may have received data but not yet processed, right?

Reply:
Data may have already reached processors/exporters (we can't guarantee that it didn't). Notice that the typical use for a readiness probe precludes that in the following sense: if the probe didn't signal "ready" no data should have reached the receivers yet, however, in environments that don't have a readiness probe there is no guarantees. OTOH in environments without readiness probe there shouldn't be the need to "add-ons" waiting on "ready" notification.

Let me know if you think a good idea to add this information there.

Reply:
You are right, my mistake, I was thinking about something else. What you have in the comment is good enough.